### ISSUE TITLE
twisted.xmlrpc.XMLRPC.render need more error handling

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2505]
|<img alt="yasusii's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| yasusii reported|
|-|-|
|Trac ID|trac#2505|
|Type|defect|
|Created|2007-03-05 08:59:26Z|

An XML-RPC message is an POST request, so the other request methods are not allowed.
xmlrpclib.loads method may cause an error and error message should be return as an XML-RPC response.

Attachments:

* [xmlrpc.py.diff.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/57b/57b99bb432afb61e5793e7c75c9a48dddf765eb6/267b64649e56a93d9e2bae7cbc1f5d247691f715.txt) (1430 bytes) - added by yasusii on 2007-03-05 08:59:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2505 2505
type__defect defect
reporter__yasusii yasusii
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1173085166000000 1173085166000000
changetime__1177869473000000 1177869473000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve set owner to @therve|
|-|-|

I'll look at it when [#2430](https://github.com/twisted/twisted/issues/2430) is merged.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

The handling of errors for xmlrpclib.loads could be useful. But I'm not sure about forcing POST requests. Do you have a precise usecase in mind?

**Comment by twisted-trac** [comment]
|<img alt="yasusii's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|yasusii commented|
|-|-|

Replying to [therve](#note_2):
> The handling of errors for xmlrpclib.loads could be useful. But I'm not sure about forcing POST requests. Do you have a precise usecase in mind?

XML-RPC Specification says:

  An XML-RPC message is an HTTP-POST request.

http://www.xmlrpc.com/spec

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

I know that. But, "be liberal about what you accept": enforcing POST doesn't seem very useful.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Ready to review in xmlrpc-error-handling-2505.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

It looks like the fix applied for [#8030](https://github.com/twisted/twisted/issues/8030) a while ago, combined with an allowedMethods declaration would be a simpler way to implement this.

Do you think it makes sense to have XMLRPC implement render_POST instead of render and to to have allowedMethods set to POST?  This should result in a NOT_ALLOWED response for anything other than POST without requiring any explicit checking for the request method in the XML-RPC code.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

I don't know/remember about allowedMethod, that is indeed an elegant way to do it.

The only thing that worry me is backward compatibility: if you override XMLRPC.render, it won't be called anymore...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

OK my objection was stupid because render will be called anyway. So ready to re-review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [20117]) Merge xmlrpc-error-handling-2505

Author: therve, yasusii
Reviewer: exarkun
Fixes [#2505](https://github.com/twisted/twisted/issues/2505)

Add more error checking in xmlrpc.XMLRPC render method, and enforce POST
requests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
